File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        //1. Initialize the pointers 
        int l = 0;
        int r = nums.length-1;
        
        //2. Handle some basic test cases 
        //a. nums.length = 1
        if(nums.length==1){
            return nums[0];
        }
        //b. case 1 - no rotation nums[r]> nums[l]
        if(nums[r]>nums[l]){
            return nums[l];
        }
        
        //3. binary search for the mid pointer 
        while(l<=r){
            //a. mid pointer 
            int mid = (int)((l+r)/2);
            
            //b. validate if its min (inflection idea)
            if(nums[mid]>nums[mid+1]){
                return nums[mid+1];
            }
            else if(nums[mid-1]>nums[mid]){
                return nums[mid];
            }
            
            //c. adjust the mid pointer 
            if(nums[mid]>nums[0]){
                l = mid+1;
            }else{
                r = mid-1;
            }
        }
        return nums[l];
    }
}

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #1. Base case 
        if len(nums) == 1:
            return nums[0]
        
        #2. Initialize the pointers for binary search 
        l = 0
        r = len(nums)-1
        
        # case 1 -> array is not rotated
        if nums[l] < nums[r]:
            return nums[l]
        
        #3. While loop 
        while l<r:
            mid = (l+r)//2
             #a. mid is too big?
            if nums[mid]>nums[0]:
                l = mid+1
            else:
                r = mid-1
            #b. compare the mid with the surrounding
            if nums[mid-1]>nums[mid]:
                return nums[mid]
            elif nums[mid]>nums[mid+1]:
                return nums[mid+1]
            
             
           
            
          
            
        return nums[l]
                
            
            
            
        